### DISCUSSION TITLE
C API feature request: Optionally allow for implementing table functions without copying data

### DISCUSSION THREAD

**Discussion by ak-coram** [duckdb/duckdb#5340] [General]
Currently the C API doesn't seem to allow for simply passing a pointer to data already in the required memory representation and instead requires copying it into pre-allocated chunks. While table functions can already be very performant even with copying data, bindings for [various languages](https://github.com/ak-coram/cl-duckdb) 😉 using the C API could also benefit from directly passing data already in memory to DuckDB for querying.

The [function stage](https://github.com/duckdb/duckdb/blob/master/test/api/capi/capi_table_functions.cpp#L38) of a table function currently receives a `duckdb_data_chunk` for the purpose of writing result columns. Using `duckdb_data_chunk_get_vector` and subsequently `duckdb_vector_get_data` on each result column (previously added during the [bind stage](https://github.com/duckdb/duckdb/blob/master/test/api/capi/capi_table_functions.cpp#L14) returns a pointer which can be used to copy values up to a size of `STANDARD_VECTOR_SIZE`. I assume that simply introducing a `duckdb_vector_set_data` function taking the vector and a pointer to the data as parameters should already be sufficient to implement table functions that can avoid copying.

I'm not familiar with how memory for vector data in table functions is managed (or whether it is reused repeatedly during query execution), but perhaps individual result columns could be flagged during the bind stage to avoid allocation of this space altogether. In any case overwriting the data pointer of vectors would probably need to at least free any previously allocated space.

Thank you for considering this!

**Comment by danz68** [comment]
I second any zero copy mechanism, yet I am not sure how viable this is for the types different than native numerical (int, double, ...). Besides, I assume the DuckDB architecture gives a high priority to one of its biggest strengths: the balanced storage mechanism (RAM/DISK). Which means that any large passed array will feel better under DuckDB umbrella, and a small one, well ... is small, right?

**Reply by ak-coram** [reply]
I agree, the utility of this for non-numeric types seems limited. I guess maybe booleans and uuids could work too. What do you think?

> Which means that any large passed array will feel better under DuckDB umbrella, and a small one, well ... is small, right?

In my [simple benchmarks](https://github.com/ak-coram/cl-duckdb/blob/main/duckdb-benchmark.lisp#L76) for my bindings using the C API I've already seen an order of magnitude of speedup compared to using the type-specific appenders just by using table functions which rely on copying. I estimate that more than half of the current runtime here is spent on copying in this specific case, seems to be definitely worth the effort to make this work.

```
Results for benchmark MEASURE-INTEGER-APPEND-SUM
-                SAMPLES  TOTAL      MINIMUM   MAXIMUM   MEDIAN    AVERAGE     DEVIATION     
REAL-TIME        100      5.057056   0.047     0.063001  0.051     0.05057056  0.0018722345  
RUN-TIME         100      5.090696   0.047467  0.063694  0.050924  0.05090696  0.0018511931  
USER-RUN-TIME    100      5.054263   0.046531  0.062851  0.050724  0.05054263  0.0018587691  
SYSTEM-RUN-TIME  100      0.036524   0         0.002042  4.0e-6    3.6524e-4   5.6624395e-4  
PAGE-FAULTS      100      0          0         0         0         0           0.0           
GC-RUN-TIME      100      16.479     0         13.138    0         0.16479     1.3455621     
BYTES-CONSED     100      321811264  3186176   3251456   3218688   3218112.8   21631.498     
EVAL-CALLS       100      0          0         0         0         0           0.0           

Results for benchmark MEASURE-STATIC-TABLE-INTEGER-VECTOR-SUM
-                SAMPLES  TOTAL     MINIMUM   MAXIMUM   MEDIAN    AVERAGE     DEVIATION     
REAL-TIME        100      0.338003  0.003     0.005     0.003     0.00338003  5.0554646e-4  
RUN-TIME         100      0.350505  0.003261  0.008833  0.003454  0.00350505  5.381717e-4   
USER-RUN-TIME    100      0.347038  0.002483  0.007852  0.003447  0.00347038  4.6242884e-4  
SYSTEM-RUN-TIME  100      0.003323  0         9.81e-4   0         3.323e-5    1.5604415e-4  
PAGE-FAULTS      100      0         0         0         0         0           0.0           
GC-RUN-TIME      100      0         0         0         0         0           0.0           
BYTES-CONSED     100      3237584   0         76032     32736     32375.84    15030.136     
EVAL-CALLS       100      0         0         0         0         0           0.0           
```

**Reply by Mytherin** [reply]
> In my [simple benchmarks](https://github.com/ak-coram/cl-duckdb/blob/main/duckdb-benchmark.lisp#L76) for my bindings using the C API I've already seen an order of magnitude of speedup compared to using the type-specific appenders just by using table functions which rely on copying. I estimate that more than half of the current runtime here is spent on copying in this specific case, seems to be definitely worth the effort to make this work.

I would do some profiling. I have eliminated copying of vectors in the past and the runtime improvement is generally closer to 5-10%. Copying data in CPU caches is incredibly cheap.

Not to say that eliminating the copies or providing this API is not worthwhile - but the benefits are not as large as you might think.

**Reply by ak-coram** [reply]
Thanks for chiming in!

> Not to say that eliminating the copies or providing this API is not worthwhile - but the benefits are not as large as you might think.

I didn't think we could cut the runtime in half here, just that there could be significant gains made :)

> I would do some profiling.

I'm not well-versed in C or C++, but if there's a proof of concept implementation in the C API, I'd be glad to do some testing with it.
